File: pkg/sdk/python/src/diverge_sdk/propagation.py
import base64
from dataclasses import dataclass, field
from enum import IntEnum
from typing import Dict

class RoutingMode(IntEnum):
    UNSPECIFIED = 0
    HEADER = 1
    SUBDOMAIN = 2
    ASYNC = 3

@dataclass
class PropagationContext:
    environment: str = ""
    routing_mode: RoutingMode = RoutingMode.UNSPECIFIED
    metadata: Dict[str, str] = field(default_factory=dict)

def _encode_varint(value: int) -> bytes:
    res = bytearray()
    while True:
        byte = value & 0x7F
        value >>= 7
        if value:
            res.append(byte | 0x80)
        else:
            res.append(byte)
            break
    return bytes(res)

def _decode_varint(buffer: bytes, offset: int) -> tuple[int, int]:
    result = 0
    shift = 0
    while True:
        if offset >= len(buffer):
            raise ValueError("Truncated varint")
        byte = buffer[offset]
        offset += 1
        result |= (byte & 0x7F) << shift
        if not (byte & 0x80):
            break
        shift += 7
    return result, offset

def _encode_string(tag: int, value: str) -> bytes:
    if not value:
        return b""
    encoded = value.encode('utf-8')
    return _encode_varint(tag) + _encode_varint(len(encoded)) + encoded

def encode_propagation_context(ctx: PropagationContext) -> str:
    out = bytearray()

    if ctx.environment:
        out.extend(_encode_string(10, ctx.environment))

    if ctx.routing_mode != RoutingMode.UNSPECIFIED:
        out.extend(_encode_varint(16))
        out.extend(_encode_varint(ctx.routing_mode.value))

    if ctx.metadata:
        for k, v in ctx.metadata.items():
            entry = bytearray()
            entry.extend(_encode_string(10, k))
            entry.extend(_encode_string(18, v))
            out.extend(_encode_varint(26))
            out.extend(_encode_varint(len(entry)))
            out.extend(entry)

    b64 = base64.standard_b64encode(out).decode('ascii').rstrip('=')
    return b64

def decode_propagation_context(encoded: str) -> PropagationContext:
    if len(encoded) > 4096:
        raise ValueError("Encoded context too large")

    pad_len = (4 - (len(encoded) % 4)) % 4
    encoded += '=' * pad_len

    try:
        buffer = base64.standard_b64decode(encoded)
    except Exception as e:
        raise ValueError(f"Invalid base64: {e}")

    ctx = PropagationContext()
    offset = 0
    length = len(buffer)

    while offset < length:
        tag, offset = _decode_varint(buffer, offset)
        field_num = tag >> 3
        wire_type = tag & 0x07

        if field_num == 1 and wire_type == 2:
            str_len, offset = _decode_varint(buffer, offset)
            ctx.environment = buffer[offset:offset+str_len].decode('utf-8')
            offset += str_len
        elif field_num == 2 and wire_type == 0:
            val, offset = _decode_varint(buffer, offset)
            try:
                ctx.routing_mode = RoutingMode(val)
            except ValueError:
                pass
        elif field_num == 3 and wire_type == 2:
            entry_len, offset = _decode_varint(buffer, offset)
            entry_end = offset + entry_len
            k = ""
            v = ""
            while offset < entry_end:
                entry_tag, offset = _decode_varint(buffer, offset)
                entry_field = entry_tag >> 3
                if entry_field == 1:
                    str_len, offset = _decode_varint(buffer, offset)
                    k = buffer[offset:offset+str_len].decode('utf-8')
                    offset += str_len
                elif entry_field == 2:
                    str_len, offset = _decode_varint(buffer, offset)
                    v = buffer[offset:offset+str_len].decode('utf-8')
                    offset += str_len
                else:
                    if (entry_tag & 7) == 0:
                        _, offset = _decode_varint(buffer, offset)
                    elif (entry_tag & 7) == 2:
                        skip_len, offset = _decode_varint(buffer, offset)
                        offset += skip_len
                    elif (entry_tag & 7) == 1:
                        offset += 8
                    elif (entry_tag & 7) == 5:
                        offset += 4
            ctx.metadata[k] = v
        else:
            if wire_type == 0:
                _, offset = _decode_varint(buffer, offset)
            elif wire_type == 2:
                skip_len, offset = _decode_varint(buffer, offset)
                offset += skip_len
            elif wire_type == 1:
                offset += 8
            elif wire_type == 5:
                offset += 4

    return ctx

File: pkg/sdk/node/src/propagation.ts
export enum RoutingMode {
  UNSPECIFIED = 0,
  HEADER = 1,
  SUBDOMAIN = 2,
  ASYNC = 3
}

export interface PropagationContext {
  environment: string;
  routing_mode: RoutingMode;
  metadata: Record<string, string>;
}

function encodeVarint(value: number): Buffer {
  const res: number[] = [];
  while (true) {
    const byte = value & 0x7F;
    value >>= 7;
    if (value) {
      res.push(byte | 0x80);
    } else {
      res.push(byte);
      break;
    }
  }
  return Buffer.from(res);
}

function decodeVarint(buffer: Buffer, offset: number): { value: number, offset: number } {
  let result = 0;
  let shift = 0;
  let bytesRead = 0;
  while (true) {
    if (offset >= buffer.length) {
      throw new Error("Truncated varint");
    }
    if (bytesRead >= 5) {
      throw new Error("Varint too long");
    }
    const byte = buffer[offset];
    offset++;
    bytesRead++;
    result = (result + ((byte & 0x7F) * Math.pow(2, shift))) >>> 0;
    if (!(byte & 0x80)) {
      break;
    }
    shift += 7;
  }
  return { value: result, offset };
}

function encodeString(tag: number, value: string): Buffer {
  if (!value) {
    return Buffer.alloc(0);
  }
  const encoded = Buffer.from(value, 'utf-8');
  return Buffer.concat([encodeVarint(tag), encodeVarint(encoded.length), encoded]);
}

export function encodePropagationContext(ctx: PropagationContext): string {
  const chunks: Buffer[] = [];

  if (ctx.environment) {
    chunks.push(encodeString(10, ctx.environment));
  }

  if (ctx.routing_mode !== RoutingMode.UNSPECIFIED && ctx.routing_mode !== undefined) {
    chunks.push(encodeVarint(16));
    chunks.push(encodeVarint(ctx.routing_mode));
  }

  if (ctx.metadata) {
    for (const [k, v] of Object.entries(ctx.metadata)) {
      const kBuf = encodeString(10, k);
      const vBuf = encodeString(18, v);
      const entry = Buffer.concat([kBuf, vBuf]);
      chunks.push(encodeVarint(26));
      chunks.push(encodeVarint(entry.length));
      chunks.push(entry);
    }
  }

  const buf = Buffer.concat(chunks);
  return buf.toString('base64').replace(/=+$/, '');
}

export function decodePropagationContext(encoded: string): PropagationContext {
  if (encoded.length > 4096) {
    throw new Error("Encoded context too large");
  }

  // Pad if necessary
  const padLen = (4 - (encoded.length % 4)) % 4;
  encoded += '='.repeat(padLen);

  const buffer = Buffer.from(encoded, 'base64');

  const ctx: PropagationContext = {
    environment: "",
    routing_mode: RoutingMode.UNSPECIFIED,
    metadata: {}
  };

  let offset = 0;

  while (offset < buffer.length) {
    let { value: tag, offset: nextOffset } = decodeVarint(buffer, offset);
    offset = nextOffset;

    const fieldNum = tag >> 3;
    const wireType = tag & 0x07;

    if (fieldNum === 1 && wireType === 2) {
      const { value: strLen, offset: next } = decodeVarint(buffer, offset);
      if (strLen < 0 || next + strLen > buffer.length) {
        throw new Error("Invalid field length");
      }
      ctx.environment = buffer.toString('utf-8', next, next + strLen);
      offset = next + strLen;
    } else if (fieldNum === 2 && wireType === 0) {
      const { value, offset: next } = decodeVarint(buffer, offset);
      ctx.routing_mode = value as RoutingMode;
      offset = next;
    } else if (fieldNum === 3 && wireType === 2) {
      const { value: entryLen, offset: next } = decodeVarint(buffer, offset);
      if (entryLen < 0 || next + entryLen > buffer.length) {
        throw new Error("Invalid field length");
      }
      offset = next;
      const entryEnd = offset + entryLen;
      let k = "";
      let v = "";
      while (offset < entryEnd) {
        const { value: entryTag, offset: eNext } = decodeVarint(buffer, offset);
        offset = eNext;
        const entryField = entryTag >> 3;

        if (entryField === 1) {
          const { value: strLen, offset: next2 } = decodeVarint(buffer, offset);
          if (strLen < 0 || next2 + strLen > buffer.length) {
            throw new Error("Invalid field length");
          }
          k = buffer.toString('utf-8', next2, next2 + strLen);
          offset = next2 + strLen;
        } else if (entryField === 2) {
          const { value: strLen, offset: next2 } = decodeVarint(buffer, offset);
          if (strLen < 0 || next2 + strLen > buffer.length) {
            throw new Error("Invalid field length");
          }
          v = buffer.toString('utf-8', next2, next2 + strLen);
          offset = next2 + strLen;
        } else {
          // Skip unknown field in entry
          const et = entryTag & 7;
          if (et === 0) {
            const { offset: next2 } = decodeVarint(buffer, offset);
            offset = next2;
          } else if (et === 2) {
            const { value: skipLen, offset: next2 } = decodeVarint(buffer, offset);
            if (skipLen < 0 || next2 + skipLen > buffer.length) {
              throw new Error("Invalid field length");
            }
            offset = next2 + skipLen;
          } else if (et === 1) {
            offset += 8;
          } else if (et === 5) {
            offset += 4;
          }
        }
      }
      ctx.metadata[k] = v;
    } else {
      // Skip unknown field
      if (wireType === 0) {
        const { offset: next } = decodeVarint(buffer, offset);
        offset = next;
      } else if (wireType === 2) {
        const { value: skipLen, offset: next } = decodeVarint(buffer, offset);
        if (skipLen < 0 || next + skipLen > buffer.length) {
          throw new Error("Invalid field length");
        }
        offset = next + skipLen;
      } else if (wireType === 1) {
        offset += 8;
      } else if (wireType === 5) {
        offset += 4;
      }
    }
  }

  return ctx;
}
